File: src/main/java/frc/robot/commands/ShootDriveCmd.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import static edu.wpi.first.units.Units.MetersPerSecond;
import static edu.wpi.first.units.Units.RadiansPerSecond;
import static edu.wpi.first.units.Units.RotationsPerSecond;

import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.DoubleSupplier;

import com.ctre.phoenix6.swerve.SwerveModule.DriveRequestType;
import com.ctre.phoenix6.swerve.SwerveRequest;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.generated.TunerConstants;
import frc.robot.subsystems.CommandSwerveDrivetrain;
import frc.robot.util.constants.FieldConstants;
import frc.robot.util.constants.SwerveConstants;

/* You should consider using the more terse Command factories API instead https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#defining-commands */
public class ShootDriveCmd extends Command {

  private CommandSwerveDrivetrain swerve;
  private DoubleSupplier translationSup;
  private DoubleSupplier strafeSup;
  private Consumer<Optional<Rotation2d>> setCurrentHeading;

  private double maxSpeed;
  private double maxAngularRate;  

  private final SwerveRequest.FieldCentricFacingAngle driveMaintainHeading;

  /** Creates a new ShootDriveCmd. */
  public ShootDriveCmd(
      CommandSwerveDrivetrain swerve,
      DoubleSupplier translationSup,
      DoubleSupplier strafeSup,
      Consumer<Optional<Rotation2d>> setCurrentHeading) {
    this.swerve = swerve;
    this.translationSup = translationSup;
    this.strafeSup = strafeSup;
    this.setCurrentHeading = setCurrentHeading;

    maxSpeed = TunerConstants.kSpeedAt12Volts.in(MetersPerSecond);
    maxAngularRate = RotationsPerSecond.of(1).in(RadiansPerSecond);

    driveMaintainHeading =
        new SwerveRequest.FieldCentricFacingAngle()
            .withDeadband(maxSpeed * 0.05)
            .withRotationalDeadband(maxAngularRate * 0.05)
            .withDriveRequestType(DriveRequestType.OpenLoopVoltage)
            .withDesaturateWheelSpeeds(true);

    /* Set the PID constants for the Maintain Heading controller */
    driveMaintainHeading.HeadingController.setPID(
        SwerveConstants.HEADING_KP, SwerveConstants.HEADING_KI, SwerveConstants.HEADING_KD);
    driveMaintainHeading.HeadingController.enableContinuousInput(-Math.PI, Math.PI);
    driveMaintainHeading.HeadingController.setTolerance(SwerveConstants.HEADING_TOLERANCE);            

    // Use addRequirements() here to declare subsystem dependencies.
    addRequirements(swerve);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {}

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    double translation = (translationSup.getAsDouble() * maxSpeed) / 3;
    double strafe = (strafeSup.getAsDouble() * maxSpeed) / 3;

    // Grab the robot's current alliance
    Optional<Alliance> alliance = DriverStation.getAlliance();

    // Choose which hub to aim to depending on alliance color
    Translation2d hubToAimTowards =
        alliance.isPresent() && (alliance.get() == Alliance.Red)
            ? FieldConstants.Hub.RED_HUB_CENTER_POSE
            : FieldConstants.Hub.BLUE_HUB_CENTER_POSE;

    // Current robot translation (x,y) in field coordinates
    var robotTranslation = swerve.getState().Pose.getTranslation();

    // Vector from robot to target hub
    var delta = hubToAimTowards.minus(robotTranslation);

    // Absolute angle in field frame pointing from robot to hub
    Rotation2d angleToPointAt = new Rotation2d(Math.atan2(delta.getY(), delta.getX()));

    setCurrentHeading.accept(Optional.of(angleToPointAt));

    swerve.setControl(
        driveMaintainHeading
            .withVelocityX(translation)
            .withVelocityY(strafe)
            .withTargetDirection(angleToPointAt)
            .withTargetRateFeedforward(0.1));

  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {}

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    return false;
  }
}

File: src/main/cpp/commands/ShootDriveCmd.cpp
#include "commands/ShootDriveCmd.h"

#include <numbers>
#include <cmath>

#include <frc/DriverStation.h>

#include "Constants.h"
#include "generated/TunerConstants.h"
#include "subsystems/CommandSwerveDrivetrain.h"

using namespace ctre::phoenix6;

ShootDriveCmd::ShootDriveCmd(
    CommandSwerveDrivetrain* swerve,
    std::function<double()> translationSup,
    std::function<double()> strafeSup,
    std::function<void(std::optional<frc::Rotation2d>)> setCurrentHeading)
    : m_swerve{swerve},
      m_translationSup{std::move(translationSup)},
      m_strafeSup{std::move(strafeSup)},
      m_setCurrentHeading{std::move(setCurrentHeading)},
      m_maxSpeed{TunerConstants::kSpeedAt12Volts.value()},
      m_maxAngularRate{2.0 * std::numbers::pi} {
  m_driveMaintainHeading.Deadband = m_maxSpeed * 0.05;
  m_driveMaintainHeading.RotationalDeadband = m_maxAngularRate * 0.05;
  m_driveMaintainHeading.DriveRequestType =
      swerve::SwerveModule::DriveRequestType::OpenLoopVoltage;
  m_driveMaintainHeading.DesaturateWheelSpeeds = true;

  m_driveMaintainHeading.HeadingController.SetPID(
      SwerveConstants::kHeadingKP, SwerveConstants::kHeadingKI,
      SwerveConstants::kHeadingKD);
  m_driveMaintainHeading.HeadingController.EnableContinuousInput(-std::numbers::pi,
                                                                  std::numbers::pi);
  m_driveMaintainHeading.HeadingController.SetTolerance(
      SwerveConstants::kHeadingTolerance);

  AddRequirements(swerve);
}

void ShootDriveCmd::Initialize() {}

void ShootDriveCmd::Execute() {
  double translation = (m_translationSup() * m_maxSpeed) / 3.0;
  double strafe = (m_strafeSup() * m_maxSpeed) / 3.0;

  auto alliance = frc::DriverStation::GetAlliance();

  frc::Translation2d hubToAimTowards =
      (alliance.has_value() && alliance.value() == frc::DriverStation::Alliance::kRed)
          ? FieldConstants::Hub::GetRedHubCenterPose()
          : FieldConstants::Hub::GetBlueHubCenterPose();

  auto robotTranslation = m_swerve->GetState().Pose.Translation();
  auto delta = hubToAimTowards - robotTranslation;

  frc::Rotation2d angleToPointAt{
      units::radian_t{std::atan2(delta.Y().value(), delta.X().value())}};

  m_setCurrentHeading(angleToPointAt);

  m_swerve->SetControl(
      m_driveMaintainHeading.WithVelocityX(units::meters_per_second_t{translation})
          .WithVelocityY(units::meters_per_second_t{strafe})
          .WithTargetDirection(angleToPointAt)
          .WithTargetRateFeedforward(units::radians_per_second_t{0.1}));
}

void ShootDriveCmd::End(bool interrupted) {}

bool ShootDriveCmd::IsFinished() {
  return false;
}
